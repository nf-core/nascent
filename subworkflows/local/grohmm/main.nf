/*
 * Run parametertuning optionally, otherwise just run transcript calling
 */

include { GROHMM_PARAMETERTUNING } from '../../../modules/local/grohmm/parametertuning/main.nf'
include { GROHMM_TRANSCRIPTCALLING } from '../../../modules/local/grohmm/transcriptcalling/main.nf'

/*
 * Note meta refers to all merged files
 */
workflow GROHMM {
    take:
    bams_bais
    gtf
    // TODO Support rerunning with a tuning file
    // tuning_file

    main:

    ch_versions = Channel.empty()

    // Run transcriptcalling eval for each tuning param
    // Should avoid a tuning file with a row for everything
    // 5..45 by 5 for UTS is what we had currently
    ch_uts = channel.fromList((params.grohmm_min_uts..params.grohmm_max_uts).step(5))
    // -100..-400 by 50 for LtProbB
    ch_ltprobb = channel.fromList((params.grohmm_min_ltprobb..params.grohmm_max_ltprobb).step(50))
    ch_bams_bais_uts_ltprobb =
        bams_bais
        .combine(ch_uts)
        .combine(ch_ltprobb)

    GROHMM_PARAMETERTUNING (
        ch_bams_bais_uts_ltprobb,
        gtf,
    )
        .tuning
        .collectFile(
            name: "${params.outdir}/transcript_identification/grohmm/tuning.csv",
            keepHeader: true,
            skip: 1,
            newLine: true,
        )
        .set { ch_tuning }
    ch_versions = ch_versions.mix(GROHMM_PARAMETERTUNING.out.versions.first())

    ch_bams_bais_tuning = bams_bais.join(ch_tuning, by: [0])


    GROHMM_TRANSCRIPTCALLING (
        ch_bams_bais_tuning,
        gtf,
    )
    ch_versions = ch_versions.mix(GROHMM_TRANSCRIPTCALLING.out.versions.first())

    emit:
    transcripts = GROHMM_TRANSCRIPTCALLING.out.transcripts
    bed = GROHMM_TRANSCRIPTCALLING.out.transcripts_bed
    td_plot = GROHMM_TRANSCRIPTCALLING.out.td_plot

    versions = ch_versions
}
