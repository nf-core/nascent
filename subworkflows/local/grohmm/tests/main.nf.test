nextflow_workflow {

    name "Test subworkflow GROHMM"
    script "../main.nf"
    workflow "GROHMM"

    test("Should run without a tuning file and multiplex") {
        when {
            params {
                outdir = "$outputDir"
                grohmm_min_uts = 5
                grohmm_max_uts = 10
                grohmm_min_ltprobb = -100
                grohmm_max_ltprobb = -150
            }
            workflow {
                """
                input[0] = Channel.of([
                    [ id: 'Sall' ],
                    [ file("https://raw.githubusercontent.com/Kraus-Lab/groHMM/master/inst/extdata/S0mR1.bam", checkIfExists: true),
                    file("https://raw.githubusercontent.com/Kraus-Lab/groHMM/master/inst/extdata/S40mR1.bam", checkIfExists: true) ],
                    [],
                ])
                input[1] = Channel.of([file(
                "https://gist.github.com/edmundmiller/c142801995689ed8d15ebcf40b2fb042/raw/eca3b955312209b5845cca084bb506d5250b3d33/hg19.chr7.refGene.gtf",
                    checkIfExists: true
                )])
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(
                    workflow.trace.tasks().size(),
                    workflow.out.transcripts,
                    workflow.out.bed,
                    // workflow.out.td_plot,
                ).match()
                },
                { assert snapshot(
                    path(workflow.out.versions.get(0)).yaml)
                 .match("versions")
                },
            )
        }
    }
}
