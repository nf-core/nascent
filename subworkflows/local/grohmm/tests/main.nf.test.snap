{
    "versions": {
        "content": [
            {
                "GROHMM:GROHMM_PARAMETERTUNING": {
                    "r-base": "4.3.3",
                    "bioconductor-grohmm": "1.39.0"
                }
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-09-24T12:58:04.996402914"
    },
    "Should run without a tuning file and multiplex": {
        "content": [
            1,
            [
                
            ],
            [
                
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-09-24T13:38:02.218159522"
    }
}